### PR TITLE
Add Create Index Error + Optional-argument integration tests for Pod + Serverless

### PR DESCRIPTION
**Pull request by austin-denoble**
## Problem
The `createIndex` integration tests are somewhat minimal and we've wanted to add thorough testing for error-cases, and optional configuration arguments for both pod and serverless indexes.

## Solution
Update `integration.controlPlane.pod.CreateDescribeAndDeleteIndexTest` and `integration.controlPlane.serverless.CreateDescribeAndDeleteIndexTest` to include additional tests validating error-conditions, and creating indexes with optional arguments.

## Type of Change
- [X] Infrastructure change (CI configs, etc) 

## Test Plan
run `gradle integrationTest` or validate the tests are run as part of the `pr` workflow in GitHub CI.

### CODE REVIEW

File: src/integration/java/io/pinecone/helpers/IndexManager.java [hunk 1]
```diff
@@ -112,7 +112,7 @@ public static Pinecone createNewIndex(Pinecone pinecone, String indexName, int d
         return pinecone;
     }
 
-    public static Index createNewIndexAndConnect(Pinecone pinecone, String indexName, int dimension, IndexMetric metric, CreateIndexRequestSpec spec) throws InterruptedException, PineconeException {
+    public static Index createNewIndexAndConnectSync(Pinecone pinecone, String indexName, int dimension, IndexMetric metric, CreateIndexRequestSpec spec) throws InterruptedException, PineconeException {
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/helpers/IndexManager.java:115]: Is the only change here the addition of the "Sync" part of the name?

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java [hunk 2]
```diff
@@ -1,51 +1,181 @@
...
+    private static final String indexName = RandomStringBuilder.build("create-index", 8);
+    private static Pinecone controlPlaneClient = new Pinecone(System.getenv("PINECONE_API_KEY"));
 
-    @Test
-    public void createAndDelete() throws InterruptedException {
-        String indexName = RandomStringBuilder.build("index-name", 8);
+    @BeforeAll
+    public static void setUp() throws InterruptedException {
+        // Create the index
         CreateIndexRequestSpecPod podSpec = new CreateIndexRequestSpecPod().environment(environment).podType("p1.x1");
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:25]: Total nit, but would it be more idiomatic to have a constant that holds the `podType` instead of hardcoding it here & then again in the `describeAndListIndex` test below? (Not sure if Java has a concept of constants, tbf)

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java [hunk 3]
```diff
@@ -1,51 +1,181 @@
...
+                .metric(IndexMetric.COSINE)
+                .dimension(10)
+                .spec(createIndexRequestSpec);
+
+        try {
+            controlPlaneClient.createIndex(createIndexRequest);
+
+            fail("Expected to throw PineconeBadRequestException");
+        } catch (PineconeBadRequestException expected) {
+            assertTrue(expected.getLocalizedMessage().contains("must consist of lower case alphanumeric characters or" +
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:96]: Nit: you might want to change the error msg here to say *what* must consist of lowercase (one word) chars... so something like "Index name must consist..." or similar

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java [hunk 4]
```diff
@@ -1,51 +1,181 @@
...
+                .metric(IndexMetric.COSINE)
+                .dimension(-1)
+                .spec(createIndexRequestSpec);
+
+        try {
+            controlPlaneClient.createIndex(createIndexRequest);
+
+            fail("Expected to throw PineconeUnmappedHttpException");
+        } catch (PineconeUnmappedHttpException expected) {
+            assertTrue(expected.getLocalizedMessage().contains("dimension: invalid value"));
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:116]: Maybe this error message could be more helpful if it said what is invalid about the user's dimension. So something like "Dimension invalid, must be integer > 0" or something

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java [hunk 5]
```diff
@@ -1,51 +1,181 @@
...
+                .metric(IndexMetric.COSINE)
+                .dimension(10)
+                .spec(createIndexRequestSpec);
+
+        try {
+            controlPlaneClient.createIndex(createIndexRequest);
+
+            fail("Expected to throw PineconeBadRequestException");
+        } catch (PineconeBadRequestException expected) {
+            assertTrue(expected.getLocalizedMessage().contains("pods"));
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:136]: This might be a totally normal Java convention, so apologies if this is obvious, but why do you have an assert stmt for bot the `"pods"` part of the message and then the rest of it (`"must be greater than 0"`)?

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java [hunk 6]
```diff
@@ -1,51 +1,181 @@
...
+                .dimension(10)
+                .spec(createIndexRequestSpec);
+
+        try {
+            controlPlaneClient.createIndex(createIndexRequest);
+
+            fail("Expected to throw PineconeBadRequestException");
+        } catch (PineconeBadRequestException expected) {
+            assertTrue(expected.getLocalizedMessage().contains("total pods must be divisible by number of shards"));
+        }
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:179]: Maybe worth adding a test to confirm an invalid similarity metric also raises an error? Just since you're doing tests for all the other index-creation params

File: src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java [hunk 7]
```diff
@@ -1,52 +1,131 @@
...
-
-    @Test
-    public void createAndDelete() throws InterruptedException {
-        String indexName = RandomStringBuilder.build("index-name", 8);
+    private static Pinecone controlPlaneClient = new Pinecone(System.getenv("PINECONE_API_KEY"));
+    private static final String indexName = RandomStringBuilder.build("create-index", 8);
+    @BeforeAll
+    public static void setUp() throws InterruptedException {
+        // Create the index
         ServerlessSpec serverlessSpec = new ServerlessSpec().cloud(ServerlessSpec.CloudEnum.AWS).region("us-west-2");
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java:23]: Might want to have a constant (or similar Java concept) for the `region` here in anticipation of us expanding to more regions?

File: src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java [hunk 8]
```diff
@@ -1,52 +1,131 @@
...
+                .metric(IndexMetric.COSINE)
+                .dimension(10)
+                .spec(createIndexRequestSpec);
+
+        try {
+            controlPlaneClient.createIndex(createIndexRequest);
+
+            fail("Expected to throw PineconeBadRequestException");
+        } catch (PineconeBadRequestException expected) {
+            assertTrue(expected.getLocalizedMessage().contains("must consist of lower case alphanumeric characters or '-'"));
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java:71]: Total nit, but technically shouldn't the "or" here be "and/or"? Currently to me it reads like the name can consist of lowercase alphanumeric chars OR just a hyphen, which isn't correct (I mean, technically you could do that I guess?)

File: src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java [hunk 9]
```diff
@@ -1,52 +1,131 @@
...
+                .metric(IndexMetric.COSINE)
+                .dimension(1)
+                .spec(createIndexRequestSpec);
+
+        try {
+            controlPlaneClient.createIndex(createIndexRequest);
+
+            fail("Expected to throw PineconeNotFoundException");
+        } catch (PineconeNotFoundException expected) {
+            assertTrue(expected.getLocalizedMessage().contains("not found"));
```
Comment (aulorbe) [inline src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java:109]: Not specific to this test, but maybe we should log a backlog issue to make the error msg for an invalid cloud more helpful than `"not found"`? Maybe something even just mentioning the resource, so like `"cloud provider not found"` and then an outlink to our documentation about current cloud providers? Not sure where to put that feedback...

Comment (aulorbe) [reply src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java:109]: (same w/region)

**Review by aulorbe — APPROVED** [review]
Just nits from me! Looks good!

**Review by rohanshah18 — APPROVED** [review]
Thanks for adding these integration tests.

File: src/integration/java/io/pinecone/helpers/IndexManager.java (hunk 1 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/helpers/IndexManager.java:115]: Yup, since it returns an `Index` I just wanted to make that clearer (as opposed to `AsyncIndex`).

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java (hunk 2 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:25]: I can set up some `static final` variables, I think that's primarily how Java deals with the idea of constants.

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java (hunk 3 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:96]: For the validations on error messages, I was trying to keep them somewhat minimal per this feedback: https://github.com/pinecone-io/pinecone-java-client/pull/81#discussion_r1527630404

So I'm using `contains` to validate that specific substrings are in the error message, which is why some of these look a little odd at first glance. My thinking was a minimal string may help prevent things failing if the error messages change, but I feel like that's somewhat unlikely and it may be beneficial to include more of the error in the assert.

I'll touch things up based on your feedback, thanks!

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java (hunk 5 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:136]: See my response above about the error assertions, I'm just checking for substrings in the larger error message.

File: src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java (hunk 9 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/integration/controlPlane/serverless/CreateDescribeListAndDeleteIndexTest.java:109]: Most of these errors are being returned from the server, so any tickets for adjusting the messages on these would need to be handled from that perspective. We're not generating the vast majority of these strings in the client.

File: src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java (hunk 6 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:179]: The argument type for `metric` is an enum: `IndexMetric`. Users are not able to pass `String` values or other values outside of this enum. I can add a test / some asserts around passing different similarity metrics though, or leaving it empty.

Comment (aulorbe) [reply src/integration/java/io/pinecone/integration/controlPlane/pod/CreateDescribeListAndDeleteIndexTest.java:176]: ah TIL!
